### PR TITLE
File io

### PR DESCRIPTION
**Pull request by jasone**
Rebased version of #71.

### CODE REVIEW

**Comment by jasone** [comment]
I left my modifications as separate changes, to be squashed before merging, to make sure the changes look okay to you, @cevans87.

File: bootstrap/src/basis/file.mli [hunk 1]
```diff
@@ -57,36 +57,36 @@ val stdout: t
...
+    permissions and and returns the resulting file or an error if the file could
+    not be opened. *)
 
 (* val of_path_hlt: ?flag:Flag.t -> ?mode:uns -> Buffer.t $-> t *)
 val of_path_hlt: ?flag:Flag.t -> ?mode:uns -> Buffer.t -> t
-(** [of_path_hlt flag mode path] opens or creates the file at [path] with
-    [flag] (default Flag.RW) Unix file permissions and [mode] (default 0o660)
-    Unix file permissions and and returns the resulting file or halts if the
-    file could not be opened. *)
+(** [of_path_hlt flag mode path] opens or creates the file at [path] with [flag]
```
Comment (cevans87) [inline bootstrap/src/basis/file.mli:69]:
```suggestion
(** [of_path_hlt ~flag ~mode path] opens or creates the file at [path] with [flag]
```

File: bootstrap/src/basis/file.mli [hunk 2]
```diff
@@ -57,36 +57,36 @@ val stdout: t
...
+val read_into_hlt: Buffer.t -> t -> unit
+(** [read_into_hlt buffer t] reads up to n bytes from [t] into the given mutable
+    [buffer], where n is the size of the buffer. Halts if bytes could not be
+    read. *)
 
 (* val read: t $-> (/t Buffer.t, Error.t) result *)
 val read: ?n:uns -> t -> (Buffer.t, Error.t) result
-(** [read n t] reads up to [n] (default 1024) bytes from [t] into a new
-    buffer and returns the buffer or an error if bytes could not be read. *)
+(** [read n t] reads up to [n] (default 1024) bytes from [t] into a new buffer
```
Comment (cevans87) [inline bootstrap/src/basis/file.mli:88]:
```suggestion
(** [read ~n t] reads up to [n] (default 1024) bytes from [t] into a new buffer
```

File: bootstrap/src/basis/file.mli [hunk 3]
```diff
@@ -170,15 +170,14 @@ module Stream : sig
 
   (* val write: file -> /t $t $-> Error.t option  *)
   val write: file -> t -> Error.t option
-  (** [write file t] takes an open [file] with write permissions and writes,
-      in order, all buffers from [t] to it. Returns an error if not all bytes
-      could be written. Returns None or an error if not all bytes could be
-      written. *)
+  (** [write file t] takes an open [file] with write permissions and writes, in
+      order, all buffers from [t] to it. Returns an error if not all bytes could
+      be written. Returns None or an error if not all bytes could be written. *)
```
Comment (cevans87) [inline bootstrap/src/basis/file.mli:175]:
```suggestion
      order, all buffers from [t] to it. Returns an error if not all bytes could
      be written. *)
```

File: bootstrap/src/basis/file.mli [hunk 4]
```diff
@@ -170,15 +170,14 @@ module Stream : sig
...
   (* val write_hlt: file -> /t $t $-> unit  *)
   val write_hlt: file -> t -> unit
-  (** [write_hlt file t] takes an open [file] with write permissions and
-      writes, in order, all buffers from [t] to it. Returns an error if not all
-      bytes could be written. Returns a unit or halts if not all bytes could
-      be written. *)
+  (** [write_hlt file t] takes an open [file] with write permissions and writes,
+      in order, all buffers from [t] to it. Returns an error if not all bytes
+      could be written. Returns a unit or halts if not all bytes could be
+      written. *)
```
Comment (cevans87) [inline bootstrap/src/basis/file.mli:182]:
```suggestion
      could be written. *)
```

**Pull request merged**